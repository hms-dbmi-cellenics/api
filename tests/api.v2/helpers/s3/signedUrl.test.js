// @ts-nocheck
const SampleFile = require('../../../../src/api.v2/model/SampleFile');
const signedUrl = require('../../../../src/api.v2/helpers/s3/signedUrl');

const AWS = require('../../../../src/utils/requireAWS');
const { NotFoundError } = require('../../../../src/utils/responses');

const {
  getSignedUrl, getSampleFileDownloadUrl, getSampleFileUploadUrls, completeMultipartUpload,
} = signedUrl;
const sampleFileInstance = new SampleFile();

jest.mock('../../../../src/api.v2/model/SampleFile');

jest.mock('../../../../src/utils/requireAWS', () => ({
  S3: jest.fn(),
}));

describe('getSignedUrl', () => {
  const signedUrlPromiseSpy = jest.fn();

  const testParams = {
    Bucket: 'test-bucket',
    Key: 'test-key',
  };

  beforeEach(() => {
    AWS.S3.mockReset();
    AWS.S3.mockImplementation(() => ({
      getSignedUrlPromise: signedUrlPromiseSpy,
    }));
  });

  it('Should call S3 signed url correctly', () => {
    getSignedUrl('getObject', testParams);
    expect(signedUrlPromiseSpy).toHaveBeenCalledWith('getObject', testParams);
  });

  it('Should add the region config if the requested url is doing upload', () => {
    const expectedConfig = {
      region: 'eu-west-1',
    };

    getSignedUrl('putObject', testParams);

    // Expect config to have region
    expect(AWS.S3).toHaveBeenCalledWith(expect.objectContaining(expectedConfig));
  });

  it('Should add the region config if the requested url is not doing upload too', () => {
    const expectedConfig = {
      region: 'eu-west-1',
    };

    getSignedUrl('getObject', testParams);

    // Expect config to have region
    expect(AWS.S3).toHaveBeenCalledWith(expect.objectContaining(expectedConfig));
  });

  it('Should throw an error if bucket is not defined', () => {
    expect(getSignedUrl('test-bucket', { Key: 'test-key' })).rejects.toThrow();
  });

  it('Should throw an error if key is not defined', () => {
    expect(getSignedUrl('test-bucket', { Bucet: 'test-bucket' })).rejects.toThrow();
  });
});

describe('getSampleFileUploadUrls', () => {
  const mockSampleFileId = 'mockSampleFileId';

  const signedUrlResponse = { signedUrls: ['signedUrl'], uploadId: 'uploadId' };

  const createMultipartUploadSpy = jest.fn();
  const getSignedUrlPromiseSpy = jest.fn();

  beforeEach(() => {
    createMultipartUploadSpy.mockReturnValue({ promise: jest.fn().mockReturnValue({ UploadId: 'uploadId' }) });
    getSignedUrlPromiseSpy.mockReturnValue('signedUrl');

    AWS.S3.mockReset();
    AWS.S3.mockImplementation(() => ({
      createMultipartUpload: createMultipartUploadSpy,
      getSignedUrlPromise: getSignedUrlPromiseSpy,
    }));
  });

  it('works correctly without metadata', async () => {
    const response = await getSampleFileUploadUrls(mockSampleFileId, {}, 1);

    expect(response).toEqual(signedUrlResponse);
    expect(createMultipartUploadSpy).toMatchSnapshot();
    expect(getSignedUrlPromiseSpy).toMatchSnapshot();
  });

  it('works correctly with metadata cellrangerVersion', async () => {
    const response = await getSampleFileUploadUrls(mockSampleFileId, { cellrangerVersion: 'v2' }, 1);

    expect(response).toEqual(signedUrlResponse);
    expect(createMultipartUploadSpy).toMatchSnapshot();
    expect(getSignedUrlPromiseSpy).toMatchSnapshot();
  });
});

describe('completeMultipartUpload', () => {
  const mockSampleFileId = 'mockSampleFileId';
  const mockParts = [];
  const mockUploadId = 'uploadId';

  const completeMultipartUploadSpy = jest.fn();

  beforeEach(() => {
    completeMultipartUploadSpy.mockReturnValue({ promise: jest.fn().mockReturnValue() });

    AWS.S3.mockReset();
    AWS.S3.mockImplementation(() => ({
      completeMultipartUpload: completeMultipartUploadSpy,
    }));
  });

  it('works correctly ', async () => {
    const response = await completeMultipartUpload(mockSampleFileId, mockParts, mockUploadId);

    expect(response).toBeUndefined();
    expect(completeMultipartUploadSpy).toMatchSnapshot();
  });
});


describe('getSampleFileDownloadUrl', () => {
  const experimentId = 'mockExperimentId';
  const sampleId = 'mockSampleId';
  const fileType = 'features10x';

  const signedUrlResponse = 'signedUrl';

  const getSignedUrlPromiseSpy = jest.fn();

  beforeEach(() => {
    getSignedUrlPromiseSpy.mockReturnValueOnce(signedUrlResponse);

    AWS.S3.mockReset();
    AWS.S3.mockImplementation(() => ({
      getSignedUrlPromise: getSignedUrlPromiseSpy,
    }));
  });

  it('works correctly', async () => {
    const files = [
      {
        id: 'id1', sampleFileType: 'matrix10x', size: 12, s3Path: '124', uploadStatus: 'uploaded', uploadedAt: '2',
      },
      {
        id: 'id0', sampleFileType: 'features10x', size: 12, s3Path: '123', uploadStatus: 'uploaded', uploadedAt: '1',
      },
    ];

    sampleFileInstance.allFilesForSample.mockImplementationOnce(() => Promise.resolve(files));

    const response = await getSampleFileDownloadUrl(experimentId, sampleId, fileType);

    expect(response).toEqual(signedUrlResponse);
    expect(getSignedUrlPromiseSpy).toMatchSnapshot();
  });

  it('Throws not found if it doesnt find a matching file', async () => {
    const files = [
      {
        id: 'id1', sampleFileType: 'matrix10x', size: 12, s3Path: '124', uploadStatus: 'uploaded', uploadedAt: '2',
      },
      {
        id: 'id0', sampleFileType: 'barcodes10x', size: 12, s3Path: '123', uploadStatus: 'uploaded', uploadedAt: '1',
      },
    ];

    sampleFileInstance.allFilesForSample.mockImplementationOnce(() => Promise.resolve(files));

    await expect(getSampleFileDownloadUrl(experimentId, sampleId, fileType)).rejects.toThrow(
      new NotFoundError(`File ${fileType} from sample ${sampleId} from experiment ${experimentId} not found`),
    );
  });
});
