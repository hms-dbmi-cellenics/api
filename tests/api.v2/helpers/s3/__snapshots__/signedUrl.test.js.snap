// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`completeMultipartUpload works correctly  1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "Bucket": "biomage-originals-test-000000000000",
        "Key": "mockSampleFileId",
        "MultipartUpload": Object {
          "Parts": Array [],
        },
        "UploadId": "uploadId",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Object {
        "promise": [MockFunction] {
          "calls": Array [
            Array [],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": undefined,
            },
          ],
        },
      },
    },
  ],
}
`;

exports[`getSampleFileDownloadUrl works correctly 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "getObject",
      Object {
        "Bucket": "biomage-originals-test-000000000000",
        "Key": "123",
        "ResponseContentDisposition": "attachment; filename=\\"features.tsv.gz\\"",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": "signedUrl",
    },
  ],
}
`;

exports[`getSampleFileUploadUrls works correctly with metadata cellrangerVersion 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "Bucket": "biomage-originals-test-000000000000",
        "Expires": 3600,
        "Key": "mockSampleFileId",
      },
    ],
    Array [
      Object {
        "Bucket": "biomage-originals-test-000000000000",
        "Expires": 3600,
        "Key": "mockSampleFileId",
        "Metadata": Object {
          "cellranger_version": "v2",
        },
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Object {
        "promise": [MockFunction] {
          "calls": Array [
            Array [],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": Object {
                "UploadId": "uploadId",
              },
            },
          ],
        },
      },
    },
    Object {
      "type": "return",
      "value": Object {
        "promise": [MockFunction] {
          "calls": Array [
            Array [],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": Object {
                "UploadId": "uploadId",
              },
            },
          ],
        },
      },
    },
  ],
}
`;

exports[`getSampleFileUploadUrls works correctly with metadata cellrangerVersion 2`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "uploadPart",
      Object {
        "Bucket": "biomage-originals-test-000000000000",
        "Expires": 3600,
        "Key": "mockSampleFileId",
        "PartNumber": 1,
        "UploadId": "uploadId",
      },
    ],
    Array [
      "uploadPart",
      Object {
        "Bucket": "biomage-originals-test-000000000000",
        "Expires": 3600,
        "Key": "mockSampleFileId",
        "Metadata": Object {
          "cellranger_version": "v2",
        },
        "PartNumber": 1,
        "UploadId": "uploadId",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": "signedUrl",
    },
    Object {
      "type": "return",
      "value": "signedUrl",
    },
  ],
}
`;

exports[`getSampleFileUploadUrls works correctly without metadata 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "Bucket": "biomage-originals-test-000000000000",
        "Expires": 3600,
        "Key": "mockSampleFileId",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": Object {
        "promise": [MockFunction] {
          "calls": Array [
            Array [],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": Object {
                "UploadId": "uploadId",
              },
            },
          ],
        },
      },
    },
  ],
}
`;

exports[`getSampleFileUploadUrls works correctly without metadata 2`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "uploadPart",
      Object {
        "Bucket": "biomage-originals-test-000000000000",
        "Expires": 3600,
        "Key": "mockSampleFileId",
        "PartNumber": 1,
        "UploadId": "uploadId",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": "signedUrl",
    },
  ],
}
`;
